app/controllers/admins_controller.rb

```diff
@@ -2,603 +2,38 @@
-                existing.save(validate: false) # Save without validations.
-              end
-              
-              rowNum +=1
-            end
-          end
-          
-          flash[:success] = "Success!"
-          
-        rescue Exception => e
-          flash[:error] = "There was an error uploading the roster
-file, most likely a duplicate email.  The exact error was: #{e} "
-          redirect_to action: "uploadRoster" and return
-        end
-        
-      else
-        # generate doIt form from the upload
-        @cuds = []
-        @currentCUDs = @course.course_user_data.all.to_a
-        @newCUDs = []
-        
-        begin
-          csv = detectAndConvertRoster(params['upload']['file'].read)
-          csv.each do |row|
-            next if (row[1].nil? || row[1].chomp.size == 0)
-            newCUD = {email: row[1].to_s,
-                      last_name: row[2].to_s.chomp(" "),
-                      first_name: row[3].to_s.chomp(" "),
-                      school: row[4].to_s.chomp(" "),
-                      major: row[5].to_s.chomp(" "),
-                      year: row[6].to_s.chomp(" "),
-                      grade_policy: row[7].to_s.chomp(" "),
-                      lecture: row[9].to_s.chomp(" "),
-                      section: row[10].to_s.chomp(" ")}
-            cud = @currentCUDs.find { |cud| 
-              cud.user and cud.user.email == newCUD[:email] 
-            }
-            if !cud then
-              newCUD[:color] = "green"
-            else
-              @currentCUDs.delete(cud) 
-            end
-            @cuds << newCUD
-          end
-        rescue CSV::MalformedCSVError => error
-          flash[:error] = "Error parsing CSV file: #{error.to_s}"
-          redirect_to :action=>"uploadRoster" and return
-        rescue Exception => e 
-          raise e
-          flash[:error] = "Error uploading the CSV file!: " +
-e.to_s() + e.backtrace().join("<br>")
-          redirect_to :action=>"uploadRoster" and return
-        end
-        
-        # drop the rest if indicated
-        if params[:upload][:dropMissing] == "1" then 
-          # We never drop instructors, remove them first
-          @currentCUDs.delete_if { |cud|
-            cud.instructor? || cud.user.administrator? || cud.course_assistant?
-          }
-          for cud in @currentCUDs do #These are the drops
-            newCUD = {email: cud.user.email,
-                      last_name: cud.user.last_name,
-                      first_name: cud.user.first_name,
-                      school: cud.school,
-                      major: cud.major,
-                      year: cud.year,
-                      grade_policy: cud.grade_policy,
-                      lecture: cud.lecture,
-                      section: cud.section,
-                      color: "red"}
-            @cuds << newCUD
-          end
-        end
+      @cuds = CourseUserDatum.select(:user_id).distinct.where(:instructor=>true)
```

Comment:
select(user_id,course_id).distinct.joins(:courses).where(courses:"end_date>Time.now", instructor:true)?
Or since you are already looping over all cuds, you could remove ones with courses with end_dates that have passed.